golang/Chapter05/model/client.go

```diff
@@ -0,0 +1,440 @@
+					pipeliner.Set(startKey, hourStart, 0)
+				}
+
+				tkey1 := uuid.NewV4().String()
+				tkey2 := uuid.NewV4().String()
+				pipeliner.ZAdd(tkey1, &redis.Z{Member: "min", Score: value})
+				pipeliner.ZAdd(tkey2, &redis.Z{Member: "max", Score: value})
+				pipeliner.ZUnionStore(destination, &redis.ZStore{Aggregate: "MIN", Keys: []string{destination, tkey1}})
+				pipeliner.ZUnionStore(destination, &redis.ZStore{Aggregate: "MAX", Keys: []string{destination, tkey2}})
+
+				pipeliner.Del(tkey1, tkey2)
+				pipeliner.ZIncrBy(destination, 1, "count")
+				pipeliner.ZIncrBy(destination, value, "sum")
+				pipeliner.ZIncrBy(destination, value*value, "sumq")
+				res, _ = pipeliner.Exec()
+				res = res[len(res)-3:]
+				return nil
+			}); err != nil {
+				log.Println("pipeline filed in UpdateStats: ", err)
+				return err
+			}
+			return nil
+		}, startKey)
+
+		if err != nil {
+			log.Println("watch filed in UpdateStats: ", err)
+			continue
+		}
+	}
+	return res
+}
+
+func (c *Client) GetStats(context, types string) map[string]float64 {
+	key := fmt.Sprintf("stats:%s:%s", context, types)
+	stats := map[string]float64{}
+	data := c.Conn.ZRangeWithScores(key, 0, -1).Val()
+	for _, v := range data {
+		stats[v.Member.(string)] = v.Score
+	}
+	stats["average"] = stats["sum"] / stats["count"]
+	numerator := stats["sumq"] - (math.Pow(stats["sum"], 2) / stats["count"])
+	count := stats["count"]
+	if count > 1 {
+		count--
+	} else {
+		count = 1
+	}
+	stats["stddev"] = math.Pow(numerator/count, 0.5)
+	return stats
+}
+
+func (c *Client) AccessTime(context string, f func()) {
+	start := time.Now().Unix()
+	f()
+	delta := time.Now().Unix() - start
+	stats := c.UpdateStats(context, "AccessTime", float64(delta), 5)
+	average := stats[1].(*redis.FloatCmd).Val() / stats[0].(*redis.FloatCmd).Val()
+
+	pipe := c.Conn.TxPipeline()
+	pipe.ZAdd("slowest:AccessTime", &redis.Z{Member:context, Score: average})
+	pipe.ZRemRangeByRank("slowest:AccessTime", 0, -101)
+	if _, err := pipe.Exec(); err != nil {
+		log.Println("pipeline err in AccessTime: ", err)
+	}
+}
+
+func (c *Client) IpToScore(ip string) int64 {
+	var score int64 = 0
+	for _, v := range strings.Split(ip, ".") {
+		n, _ := strconv.ParseInt(v, 10, 0)
+		score = score*256 + n
+	}
+	return score
+}
+
+func (c *Client) ImportIpsToRedis(filename string) {
+	res := utils.CSVReader(filename)
+	pipe := c.Conn.Pipeline()
+	for count, row := range res {
+		var (
+			startIp string
+			resIP   int64
+		)
+		if len(row) == 0 {
+			startIp = ""
+		} else {
+			startIp = row[0]
+		}
+		if strings.Contains(strings.ToLower(startIp), "i") {
+			continue
+		}
+		if strings.Contains(startIp, ".") {
+			resIP = c.IpToScore(startIp)
+		} else {
+			var err error
+			resIP, err = strconv.ParseInt(startIp, 10, 64)
+			if err != nil {
+				continue
+			}
+		}
+		cityID := row[2] + "_" + strconv.Itoa(count)
+		pipe.ZAdd("ip2cityid:", &redis.Z{Member: cityID, Score: float64(resIP)})
```

Comment:
Good improvement! I should have been using pipelines here, given the fact that we're importing many thousands of lines of information (for the full dataset). I'm not sure why I didn't when I wrote the chapter.